lib/sync-DataSetModel.js

```diff
@@ -107,66 +133,147 @@ var self = {
     self.getDataset(dataset_id, function (err, dataset) {
       if (err) return cb(err);
+
+      var done = self.wrapHandlerCallback(
+        'read',
+        {
+          dataset_id: dataset_id,
+          uid: uid,
+          meta_data: meta_data
+        },
+        cb
+      );
+
       if (dataset.readHandler) {
-        dataset.readHandler(dataset_id, uid, cb, meta_data);
+        dataset.readHandler(dataset_id, uid, done, meta_data);
       }
       else {
-        self.globalReadHandler(dataset_id, uid, cb, meta_data);
+        self.globalReadHandler(dataset_id, uid, done, meta_data);
       }
     });
   },
 
   doUpdateHandler: function (dataset_id, uid, data, cb, meta_data) {
     self.getDataset(dataset_id, function (err, dataset) {
       if (err) return cb(err);
+
+      var done = self.wrapHandlerCallback(
+        'update',
+        {
+          dataset_id: dataset_id,
+          uid: uid,
+          data: data,
+          meta_data: meta_data
+        },
+        cb
+      );
+
       if (dataset.updateHandler) {
-        dataset.updateHandler(dataset_id, uid, data, cb, meta_data);
+        dataset.updateHandler(dataset_id, uid, data, done, meta_data);
       }
       else {
-        self.globalUpdateHandler(dataset_id, uid, data, cb, meta_data);
+        self.globalUpdateHandler(dataset_id, uid, data, done, meta_data);
       }
     });
   },
 
   doDeleteHandler: function (dataset_id, uid, cb, meta_data) {
     self.getDataset(dataset_id, function (err, dataset) {
       if (err) return cb(err);
+
+      var done = self.wrapHandlerCallback(
+        'delete',
+        {
+          dataset_id: dataset_id,
+          uid: uid,
+          meta_data: meta_data
+        },
+        cb
+      );
+
       if (dataset.deleteHandler) {
-        dataset.deleteHandler(dataset_id, uid, cb, meta_data);
+        dataset.deleteHandler(dataset_id, uid, done, meta_data);
       }
       else {
-        self.globalDeleteHandler(dataset_id, uid, cb, meta_data);
+        self.globalDeleteHandler(dataset_id, uid, done, meta_data);
       }
     });
   },
 
   doCollisionHandler: function (dataset_id, hash, timestamp, uid, pre, post, meta_data) {
     self.getDataset(dataset_id, function (err, dataset) {
       if (err) return cb(err);
+
+      self.events.emit('collision-handle', {
+        dataset_id: dataset_id,
+        hash: hash,
+        timestamp: timestamp,
+        uid: uid,
+        pre: pre,
+        post: post,
+        meta_data: meta_data
+      });
+
+      // var done = wrapHandlerCallback(
+      //   'collision-handle',
+      //   {
+      //     dataset_id: dataset_id,
+      //     hash: hash,
+      //     timestamp: timestamp,
+      //     uid: uid,
+      //     pre: pre,
+      //     post: post,
+      //     meta_data: meta_data
+      //   },
+      //   cb
+      // );
+
       if (dataset.collisionHandler) {
         dataset.collisionHandler(dataset_id, hash, timestamp, uid, pre, post, meta_data);
```

Comment:
I think we should change the `collisionHandler` here to accept a callback function if it's defined to keep the API consistent.

@evanshortiss WDYT?